### DISCUSSION TITLE
Is it possible to export previews if originals go missing?

### DISCUSSION THREAD

**Discussion by vsee** [RhetTbull/osxphotos#445] [Q&A] [answered]
Hi and thanks for sharing this extensive tool!

I recently lost a bunch of my photos due to a cyberattack on my local network storage. Thankfully, I had a backup of most of them but still a large chunk are missing. I had setup the Photos app so that I would have all pictures on the external network storage being referenced from my Mac. Now that the originals are missing on the storage, the Photos app gives me the expected warning that it can no longer find them. Since I can still see the missing pictures and all their metadata, there must be some probably smaller version saved locally on my mac. I was trying to find a way to export those preview pictures at least which is better than nothing but so far I have not found a way to do that. I was wondering if this tool would be able to do that by extracting the preview pictures from the database or where ever they might be stored locally.

thanks!
Volker

**Comment by RhetTbull** [comment, marked as answer]
Hi @vsee.  Sorry to hear about your unfortunate situation!  Currently it is not possible to export preview images with osxphotos though it is on my roadmap (#50).  What version of MacOS are you running?

**Reply by RhetTbull** [reply]
I've just updated the code to be able to extract the paths for the preview images for Catalina and Big Sur. With this change I can now add an option like `export-preview-if-missing` to osxphotos.  

What is you actually want to do?  Do you want to just export the photos for backup purposes?  Or are you interested in "repairing" the library?  If the latter, I have another library called [PhotoScript](https://github.com/RhetTbull/PhotoScript) that can modify Photos (for example, import new photos).  With this it would be possible to write a simple script that finds all photos that are missing the original images, gets the preview image and re-imports the preview image as a new image into Photos.  It could even copy all metadata, place the new images in the same albums as the original, etc.

**Reply by vsee** [reply]
Hi @RhetTbull, thanks for you fast reply! An option like ```export-preview-if-missing``` sounds exactly like what I would need! I am running Big Sur 11.3.

I can recover most of my media files from an external backup and would be able to put them back into the original folder on the network storage. I supposed, Photos should then be able to find the original for those again without problem. The ones that are missing entirely I wanted to export as previews and put them back into the original location on the NAS as well so Photos could pick them up. I am also using other media libraries such as Plex to share my pictures through the net, therefore I wanted the exported version on the network storage. Photoscript sounds like it could automate that last step which I would have done manually, i.e. export previews with oxphotos, sort according to metadata into folders, move them back to network storage and tell Photos to find them there or reimport as new pictures.

As an aside:
I was actually able to use a recovery tool to get some of those missing pictures back in their original form. But they are now entirely without metadata and have random filenames in a single folder with tens of thousands of others that I don't need. So I would need something that can find pictures based on similarity in a file system and reimport them. I don't suppose PhotoScript and oxphotos can do that yet ;)

**Reply by vsee** [reply]
I will have a look at ```path-derivatives``` for now and see how far I get with a script that can pick out the largest ones. This is already a massive help, thanks so much!

**Reply by vsee** [reply]
Yes, this did the trick! Using a small python script and osxphotos I could recover the previews for all my missing pictures. Thanks again for your help! Really made my day :)

**Reply by vsee** [reply]
It is not very polished and paths are hardcoded but in case anyone finds this as useful as I did, here is my script:

```python
import osxphotos
import sys
from pathlib import Path
import shutil
import os

def copyFile(src, tdir, tname):
    if not os.path.exists(tdir):
        print("ERROR: target path does not exist: " + tdir)
        return False

    target = os.path.join(tdir,tname)

    try:
        shutil.copy2(src, target)
        print(src + " --> " + target)
    except shutil.SameFileError:
        print("ERROR: Source " + src + " and destination " + target + " are the same.")
        return False
    except IOError:
        print("ERROR: copy source " + src + " to destination " + target + " failed.")
        return False

    return True

def copy_largest(photo, derivatives, target_dir):
	largest = None
	largest_size = -1
	for d in derivatives:
		d_size = Path(d).stat().st_size 
		if not largest or d_size > largest_size:
			largest_size = d_size
			largest = d

	if not copyFile(largest, target_dir, photo.filename):
		print("WARNING: copying photo " + photo.filename + " failed!")
		return False
	return True
			


def main():
	print("Loading Database ...")
	photosdb = osxphotos.PhotosDB("/Users/vseeker/Pictures/Photos Library.photoslibrary")
	albums = photosdb.album_info

	# find album with lost pictures called lost
	lost_a = None
	for a in albums:
		if a.title == "lost":
			lost_a = a
			break

	if not lost_a:
		print("ERROR: Album titled lost not found.")
		sys.exit(1)
	else:
		print("Album with " + str(len(lost_a)) + " lost pictures found")

	target_dir = "/Users/vseeker/workspace/photos_script/recovery"
	print("Creating target recovery path at: " + target_dir)
	Path(target_dir).mkdir(parents=False, exist_ok=True)

	# go through derivative paths and find the largest
	noderiv = 0
	recovered = 0
	failed = 0
	for p in lost_a.photos:
		if len(p.path_derivatives) == 0:
			noderiv += 1
		else:
			if copy_largest(p, p.path_derivatives, target_dir):
				recovered += 1
			else:
				failed += 1

	print("No derivatives found for " + str(noderiv) + " lost photos.")
	print("Recovery failed for " + str(failed) + " lost photos.")
	print("Recovery successful for: " + str(recovered) + " lost photos")

if __name__ == '__main__':
	main()
```

**Reply by RhetTbull** [reply]
Great! I'm glad you were able to recover the photos.  I've opened an issue (#446) to add an `--export-preview-if-missing` option to osxphotos.  For your recovery script, a couple things to consider:

1. The preview will have extension .jpeg but if the original was different (e.g. .HEIC, .TIFF, etc) simply copying the preview to the right location won't work as Photos won't find it.  Though not clean, you could take advantage of what I consider a bug in Photos and simply rename the file with the original extension (see [PhotoInfo.original_filename](https://github.com/RhetTbull/osxphotos#original_filename)) -- Photos will properly display the photo even though the extension is wrong.  Unfortunately when this happens, osxphotos will use the original extension when exporting and any other program you use to view/edit the photos will likely complain.
2. For video files there are sometimes two previews -- one is a still image .jpeg used as the "cover" preview in Photos and one is a preview of the video as a .mov file.  If you have videos in your collection you'll want to be sure you're exporting the .mov preview (which may or may not exist).

>  So I would need something that can find pictures based on similarity in a file system and reimport them. I don't suppose PhotoScript and oxphotos can do that yet ;)

No, but as you know python, there is the [ImageHash](https://pypi.org/project/ImageHash/) library which does "visual hashing".  You could combine this with osxphotos to do something like this: compute the hash of all your recovered images, then go through your `lost` album (or just look for any photo where `photo.path is None`), get the preview image (`path_derivatives`) and compute the hash then if you find a match, move the recovered image back to the original location.  One thing you'll need to solve is finding the original location to restore to.  

`PhotoInfo.original_filename` will get you the original filename as it was imported to Photos but `Photo.path` will return None because it'll detect the original is missing.  To solve this, you can use the internals of osxphotos to get the actual directory where the photo was stored on disk using `photo._info["directory"]` (all the data about a photo is stored internally in the PhotoInfo object in a dict called `_info`):

```pycon
>>> photo = photosdb.get_photo("8051550D-94C8-401E-9829-12BBA760425E")
>>> photo._info["directory"]
'/Users/rhet/Dropbox/Screenshots'
>>> photo.original_filename
'Screen Shot 2021-05-02 at 10.10.26 PM.png'
```

If `photo._info["directory"]` does not start with a `/` then the file is stored in the library and you'll need to find the path a different way. You can reference the source for PhotoInfo.path:

https://github.com/RhetTbull/osxphotos/blob/98b84c17f176e7aa07a7d45d2c0cff3def929f18/osxphotos/photoinfo/photoinfo.py#L196-L209

I've not used ImageHash so I don't know how well it works when the two images are different sizes as they will be in this case but I'm guessing there's a way to work around this.  If this works it certainly beats doing it by hand!

**Reply by RhetTbull** [reply]
Not sure why I didn't think of this when implementing path_derivatives, but in [v0.42.20](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.20) I've fixed `path_derivatives` to return results in reverse sorted order by size.  That is, `photo.path_derivatives[0]` will always be the largest preview image available.  Of course, you'll still need to check for .mov vs .jpeg for videos and Live Photos.

**Reply by vsee** [reply]
re 1. that is good advice! I will probably just remove the missing pictures from my library and reimport the previews I recovered.
re 2. I did not actually save any videos in the Photos library.

ImageHash sounds like something that could work, I will have a look at that, thanks.

Ah, sorting by size simplifies this a bit :)

**Comment by RhetTbull** [comment]
Hi @vsee I've added a `--preview-if-missing` option to [v0.42.58](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.58)
